### PR TITLE
VC - Enable display price formula

### LINKED ISSUE #13039: VC: Enable display price formula in configuration
Take into account that there are `ConfiguratorAttributeSingleSelectionBaseComponent` and `ConfiguratorAttributeMultiSelectionBaseComponent` base components that 'contain' `extractPriceFormulaParameters` method that prepares price formula.

**Wireframes:**
[M_CPS_Configuration_forms](https://4prsat.axshare.com/#id=jhn1qb&p=m_cps_configuation_forms&g=1)
[T_CPS_Configuration_forms](https://4prsat.axshare.com/#id=4vlcax&p=t_cps_configuration_forms&g=1)
[D_CPS_Configuration_forms](https://4prsat.axshare.com/#id=fr0uxa&p=d_cps_configuration_forms&g=1)
[W_CPS_Configuration_forms](https://4prsat.axshare.com/#id=zq2caa&p=w_cps_configuration_forms&g=1)

Price formula should be enabled for the following attribute types:
- [x] Drop-down listbox
- [x]  Single-select radio button
- [ ]  Input Field (not possible)
- [ ]  Image (VC / CPS)
- [x]  Multi-select checkbox

**Done criteria:**
- [x] styling,
- [ ] corresponding unit tests

### PR DESCRIPTION
**Pull request by Larisa-Staroverova**

### CODE REVIEW

File: feature-libs/product-configurator/rulebased/core/state/reducers/configurator.reducer.ts [hunk 1]
```diff
@@ -181,14 +187,37 @@ function takeOverChanges(
...
+  };
+  return result;
+}
+
+function takeOverPriceSupplementsChanges(
+  action: ConfiguratorActions.UpdatePriceSummarySuccess,
+  state: Configurator.Configuration
+): Configurator.Configuration {
+  const content = { ...action.payload };
+  const groups = ConfiguratorStateUtils.deepCopy(state.groups);
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/core/state/reducers/configurator.reducer.ts]: That we need to discuss

Comment (Platonn) [reply feature-libs/product-configurator/rulebased/core/state/reducers/configurator.reducer.ts]: We should implement differently `ConfiguratorStateUtils.updateValuePrices`. Instead of being void function, it should return a new, fresh object. Moreover it should create a fresh object on whole levels of the change, according to the [immutable update pattern](https://redux.js.org/usage/structuring-reducers/immutable-update-patterns). Thanks to that, only the updated branches of the state are copied, which is optimal. 

Deep copy of the whole object is not justified IMO, becasue:
  - we're also coping branches of the state that are not updated
  - it introduces a custom generic code for deep-copying objects, which requires additional good unit tests coverage 
  - btw. IMO it’s not the business of ConfiguratorStateUtils to deep copy objects (it's rather a generic util) 

Explanation, why we got a runtime error `ERROR TypeError: Cannot add property valuePrice, object is not extensible`:
To help developers, ngrx in [dev](https://github.com/ngrx/platform/blob/master/modules/store/src/runtime_checks.ts#L21) mode [freezes](https://github.com/ngrx/platform/blob/master/modules/store/src/meta-reducers/immutability_reducer.ts#L18) in runtime the objects of the state and actions -> to help identify accidental, unwanted mutations (which could lead in some cases to selectors' Observables not emitting a new value). FYI the freezing behavior can be turned off [by the configuration]( https://ngrx.io/guide/store/configuration/runtime-checks), but I don't recommend it.

File: feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        (value) => value === 'NOT EXISTING',
+        (value) => {
+          return value;
+        }
+      );
+      expect(mergedValues).toBeUndefined();
+    });
+  });
+
+  describe('mergeGroupsWithSupplements', () => {
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts:151]: I would add a test with empty attributeSupplements list.
Then we have tested both the case that the group list is empty and the supplements list is empty

File: feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+      value: 12,
+    };
+    let values: Configurator.Value[] = [{ valueCode: code }];
+    const valueSupplement: Configurator.ValueSupplement = {
+      attributeValueKey: code,
+      priceValue: priceDetails,
+      obsoletePriceValue: priceDetails,
+    };
+
+    it('should create new array in case predicate matches', () => {
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts:113]: Correct me if I am wrong, but I have not seen a test that checks that the projection is done correctly and for the correct entry.
Please add this test.

File: feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.adapter.ts [hunk 4]
```diff
@@ -230,22 +230,23 @@ export class VariantConfiguratorOccAdapter
...
-        VARIANT_CONFIGURATOR_PRICE_SUMMARY_NORMALIZER
-      ),
-      map((pricingResult) => {
+      this.converterService.pipeable(VARIANT_CONFIGURATOR_PRICE_NORMALIZER),
+      map((configResult) => {
         const result: Configurator.Configuration = {
           ...configuration,
-          priceSummary: pricingResult,
+          priceSummary: configResult.priceSummary,
+          priceSupplements: configResult.priceSupplements,
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.adapter.ts:243]: The method is called readPriceSummary, but the priceSupplements are set here as well. This is kind of misleading. We cannot rename the method can we?

File: feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.converters.ts [hunk 5]
```diff
@@ -11,10 +11,15 @@ export const VARIANT_CONFIGURATOR_SERIALIZER = new InjectionToken<
   Converter<Configurator.Configuration, OccConfigurator.Configuration>
 >('VariantConfiguratorSerializer');
 
+//TODO: pricing
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.converters.ts]: Please elaborate what this TODO is used for.

File: feature-libs/product-configurator/rulebased/occ/variant/converters/occ-configurator-variant-price-normalizer.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import { ConfiguratorModelUtils } from '@spartacus/product-configurator/common';
+import { OccConfigurator } from '../variant-configurator-occ.models';
+import { Configurator } from './../../../core/model/configurator.model';
+
+@Injectable({ providedIn: 'root' })
+export class OccConfiguratorVariantPriceNormalizer
+  implements Converter<OccConfigurator.Prices, Configurator.Configuration> {
+  convert(
+    source: OccConfigurator.Prices,
+    target?: Configurator.Configuration
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/occ/variant/converters/occ-configurator-variant-price-normalizer.ts:12]: Why are we converting to a configuration object which we do not use any further?

Comment (schuppell) [reply feature-libs/product-configurator/rulebased/occ/variant/converters/occ-configurator-variant-price-normalizer.ts:12]: The config object is only used to read the price summary and the price supplements

File: feature-libs/product-configurator/rulebased/occ/variant/converters/occ-configurator-variant-price-normalizer.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      this.convertAttributeSupplements(attr, priceSupplements);
+    });
+    const resultTarget: Configurator.Configuration = {
+      ...target,
+      configId: source.configId,
+      groups: [],
+      flatGroups: [],
+      owner: ConfiguratorModelUtils.createInitialOwner(),
+      interactionState: {},
+      priceSummary: source?.priceSummary,
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/occ/variant/converters/occ-configurator-variant-price-normalizer.ts:25]: We should call the price summary converter here: 
```
priceSummary: this.converterService.convert(
        prices,
        VARIANT_CONFIGURATOR_PRICE_SUMMARY_NORMALIZER
      ),
```

File: feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.models.ts [hunk 8]
```diff
@@ -19,12 +19,25 @@ export namespace OccConfigurator {
...
   export interface Prices {
-    configId?: string;
+    configId: string;
+    attributes?: Supplements[];
     pricingError?: boolean;
     showDeltaPrices?: boolean;
     priceSummary?: PriceSummary;
   }
 
+  export interface Supplements {
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.models.ts:29]: I would suggest to call the interface AttributeSupplements to make it consistent with ValueSupplements below.

File: feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.models.ts [hunk 9]
```diff
@@ -19,12 +19,25 @@ export namespace OccConfigurator {
...
+    csticUiKey: string;
+    selectedValues: string[];
+    priceSupplements: ValueSupplements[];
+  }
+
+  export interface ValueSupplements {
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.models.ts:35]: ValueSupplements -> ValueSupplement
Is it correct that we use plural here?
The same is valid for Supplements above.

**Review by schuppell — CHANGES_REQUESTED** [review]
See comments

File: feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts (hunk 2 above)
Comment (ChristophHi) [reply feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts:151]: In this case an error will be thrown. Was already covered, but yes, for clarity I added this test

File: feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.adapter.ts (hunk 4 above)
Comment (ChristophHi) [reply feature-libs/product-configurator/rulebased/occ/variant/variant-configurator-occ.adapter.ts:243]: You are right! But as you suppose we think the effort to rename the method is too high. We want to keep 'readPriceSummary' even if a bit misleading now. We can think of migrating it in a future major release

File: feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts (hunk 3 above)
Comment (ChristophHi) [reply feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts:113]: You are right. This is covered by the tests of the main method, but I will add a test for clarity reasons

Comment (ChristophHi) [reply feature-libs/product-configurator/rulebased/core/state/configurator-state-utils.spec.ts:113]: Done

File: feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.module.ts [hunk 10]
```diff
@@ -4,6 +4,7 @@ import { FormsModule, ReactiveFormsModule } from '@angular/forms';
 import { I18nModule } from '@spartacus/core';
 import { KeyboardFocusModule } from '@spartacus/storefront';
 import { ConfiguratorAttributeCheckBoxComponent } from './configurator-attribute-checkbox.component';
+import { ConfiguratorPriceModule } from '../../../price/configurator-price.module';
```
Comment (steinsebastian) [inline feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.module.ts:7]: Why is ConfiguratorPriceModule imported? I can't see any usage.

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.module.ts:7]: It is imported to be used within the HTML component

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.module.ts:7]: Otherwise the price component is unknown

Comment (steinsebastian) [reply feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.module.ts:7]: In configurator-attribute-checkbox.component.html I don't see any usage.

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.module.ts:7]: I see that the price component was accidentally removed from the HTML

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.module.ts:7]: It should be added, otherwise the price won't be displayed

**Review by steinsebastian — DISMISSED** [review]

**Review by schuppell — APPROVED** [review]
